boa/src/builtins/object/internal_methods.rs

```diff
@@ -361,21 +363,30 @@ impl Object {
         }
     }
 
-    /// Inserts a field in the object `properties` without checking if it's writable.
+    /// Inserts a field in the object `properties`, given an attribute without checking if it's writable.
     ///
     /// If a field was already in the object with the same name that a `Some` is returned
     /// with that field, otherwise None is retuned.
     #[inline]
-    pub(crate) fn insert_field<K>(&mut self, key: K, value: Value) -> Option<Property>
+    pub(crate) fn insert_property<K, V>(
+        &mut self,
+        key: K,
+        value: V,
+        attribute: Attribute,
+    ) -> Option<Property>
     where
         K: Into<PropertyKey>,
+        V: Into<Value>,
     {
-        self.insert_property(
-            key.into(),
-            Property::data_descriptor(
-                value,
-                Attribute::WRITABLE | Attribute::ENUMERABLE | Attribute::CONFIGURABLE,
-            ),
-        )
+        let property = Property::data_descriptor(value.into(), attribute);
+        match key.into() {
+            PropertyKey::Index(index) => self.indexed_properties.insert(index, property),
+            PropertyKey::String(ref string) => {
+                self.string_properties.insert(string.clone(), property)
+            }
+            PropertyKey::Symbol(ref symbol) => {
+                self.symbol_properties.insert(symbol.clone(), property)
+            }
+        }
     }
```

Comment:
If insert retruns an `Option<Property>` we could do:
```rust
let property = Property::data_descriptor(value.into(), attribute);
self.insert(key.into(), property)
```
This would eliminate some code duplication.